### ISSUE TITLE
Feature Request: create possibility to enforce CS1718 ("Comparison made to same variable, ...")

### ISSUE THREAD

**Issue by TobiasKnauss** [dotnet/roslyn#43789]
When comparing a variable with itself, it creates a compiler warning CS1718 ("Comparison made to same variable, ...").
This warning should also be given when the same variable is given as both arguments to the Equals() function.
Also, I would like to have a possibility to enforce CS1718 on my own Equals() overloads.

**Comment by jcouv** [comment]
The biggest hurdle here is that you're proposing a breaking change (existing projects would no longer compile). We would recommend implementing an analyzer for adding such a warning.

If such a change were to be made, what are you proposing to consider "the Equals() function"?
The `object.Equals(object, object)` method?
Any `bool Equals(...)` instance method with identical parameter types?
What about methods called "Same"?

Tagging @gafter to confirm, but I don't think this proposal would qualify as a warning analyzer/waves candidate.

**Comment by gafter** [comment]
I agree. I don't think this is something we'd likely do in the compiler.

**Comment by TobiasKnauss** [comment]
I see that it would be a breaking change, which we have to avoid of course. But I don't like to discard the whole idea yet.
My original intention to write this request was the second question: I would like to have the possibility, e.g. via attribute, to enforce this warning on my own functions.
And regarding existing functions: this could be a warning that is switched of by default. IIRC, similarly to the warning on the new (non)-nullable reference types, which also was kind of breaking change.

**Comment by gafter** [comment]
@TobiasKnauss We very much encourage you to deploy an analyzer that does this for you.

**Comment by TobiasKnauss** [comment]
@gafter This might be difficult:
- the compiler warning may be switched on or off. This also applies to the analyzer warning.
- the compiler warning may be disabled and restored at specific code sections by `#pragma warning ...`. The analyzer warning may not. But this will be needed because there are situations (e.g. in unit tests) where you intentionally put the same object into both sides of  `equals()`.

**Comment by gafter** [comment]
Reopening for the analyzer team to help you with how to suppress diagnostics produced by analyzers.

**Comment by mavasani** [comment]
@jinujoseph Can you transfer this to analyzers repo?
